Refine GHC deprecation policy (#3438)
haskell/haskell-language-server: docs/support/ghc-version-support.md

## GHC version deprecation policy

### Base policy

This is the static part of the policy that can be checked by a machine.

#### Major versions

> A major GHC version is a "legacy" version if it is 3 or more major versions behind the latest GHC version that is

Reviewer: are we like switching to the “oldest popular version” as min(stackage_lts, ghcup_recommended). As in https://discourse.haskell.org/t/call-for-opinions-shortening-the-hls-support-window/9641/16?u=soulomoon

Reviewer 2: Yes, we are! Didn't have time to add this section yet :)

Reviewer 2: Updated!

HLS will support all non-legacy major versions of GHC.

#### Minor versions

For the latest supported major GHC version we will support at least 2 minor versions.

For the rest of the supported major GHC versions, we will support at least the latest minor version in Stackage LTS (so 1 minor version).

> ### Extended policy

Reviewer 3: I think the base/extended distinction is confusing and you could just drop these headers. It's fairly clear that the later sections talk about exceptions to the rules.

Author: Hm, I think we're having a slightly different perception. 

I'm not seeing it as an exception rule that can be applied if someone speaks up about it.

Someone actually updating the supported GHC versions is **required** to check/consided both.

Reviewer 3: Okay, yes, I agree with that. Then maybe we just want to add after line 81: "HLS may continue to support legacy GHC versions, see the section below on 'Supporting legacy GHC versions'" or something?

This is the part of the policy that needs evaluation by a human and possibly followed
by a discussion.

#### Ecosystem factors

To establish and apply the policy we also take into account:

> - Completeness: support includes all plugins and features

Reviewer 3: This is already included above

Author: I don't understand this comment

Author: Included where?

Do you mean 

> Support status (see the support policy below for more details):
> - "initial support": support for this GHC version is underway, but it is not ready to be released yet
> - "basic support": this GHC version is currently actively supported, and all [tier 1 plugins](./plugin-support.md) work
> - "full support": this GHC version is currently actively supported, and most [tier 2 plugins](./plugin-support.md) work
> - "deprecated": this GHC version was supported in the past, but is now deprecated

?

Reviewer 2: I rephrased this to make sure it is clear we are referring to the HLS support status.

Also, I made clear that we aim to list *all* factors.

> - The most recent [stackage](https://www.stackage.org/) LTS snapshot

Reviewer 3: ditto

- The GHC versions used in the most popular [linux distributions](https://repology.org/project/ghc/versions)
- The reliability of different ghc versions on the major operating systems (Linux, Windows, MacOS)

> - The [Haskell Survey results](https://taylor.fausak.me/2022/11/18/haskell-survey-results/#s2q4)

Reviewer 3: I think these three points are the interesting ones. I'd move them into the next section where they're used.

> ### Supporting a GHC version beyond its shelf life

Reviewer 3: ```suggestion
### Supporting a GHC version beyond normal deprecation time
```

> In cases where the base policy demands a deprecation, but ecosystem factors
> suggest that it's still widely used (e.g. last [Haskell Survey results](https://taylor.fausak.me/2022/11/18/haskell-survey-results/#s2q4)), the deprecation should be suspended for the next release and the situation be re-evaluated for the release after that.

Reviewer 3: ```suggestion
Sometimes the policy suggests that we should deprecate a version of GHC, but doing so would be particularly harmful for our users because it remains very widely used. In cases like this we can decide to keep supporting that GHC version, but this decision should be revisited each time we do a release.

To determine whether we should keep supporting a GHC version we can use 
- The GHC versions used in the most popular [linux distributions](https://repology.org/project/ghc/versions)
- The [Haskell Survey results](https://taylor.fausak.me/2022/11/18/haskell-survey-results/#s2q4)
- Anecdotal evidence about he reliability of different GHC versions on the major operating systems (Linux, Windows, MacOS), e.g. noting the presence of serious bugs
```

Author: I think this is mixing up things. Ecosystem factors should be its own section and be referenced from this section explaining the **process** (not criteria) of supporting an older GHC version.

Reviewer 2: I agree and think this separation helps the reader understanding what is going on.

> In case of maintenance burden concerns, the following process should be followed:

Reviewer 3: ```suggestion
When we decide to keep on an old version, we should track it as follows:
```

1. open a ticket on HLS issue tracker wrt discussing to deprecate said GHC version
    - explain the reason the GHC version wasn't deprecated (context)
    - explain the maintenance burden it causes (reason)
    - evaluate whether it impacts the next HLS release (impact)
2. discuss whether ecosystem factors changed
    - e.g. if Haskell Survey results show that 25% or more of users are still on the GHC version in question, then dropping should be avoided
3. if dropping is still undesired, but maintenance burden is also high, then set out a call-for-help and contact HF for additional funding to support this GHC version

> 4. if no help or funding was received within 2 releases (say, e.g. 3-6 months), then drop the version regardless

Reviewer 3: I'm fine with this but we could also leave it out and just experiment with various ways of asking for help before we write it down

Author: True
